host/xtest/pkcs11_1000.c

```diff
@@ -1463,3 +1483,407 @@ static void xtest_pkcs11_test_1007(ADBG_Case_t *c)
+		if (test->in != NULL) {
+			rv = C_SignInit(session, test->mechanism, key_handle);
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err_destr_obj;
+
+			rv = C_SignUpdate(session,
+					  (void *)test->in, test->in_len);
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err_destr_obj;
+
+			/* Test too short buffer case */
+			out_size = 1;
+			rv = C_SignFinal(session, out, &out_size);
+			if (!ADBG_EXPECT_CK_RESULT(c, CKR_BUFFER_TOO_SMALL, rv))
+				goto err_destr_obj;
+
+			/*
+			 * Test NULL buffer case with size as 0
+			 * to get the out_size
+			 */
+			out_size = 0;
+			rv = C_SignFinal(session, NULL, &out_size);
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err_destr_obj;
+
+			/* Get to full output */
+			memset(out, 0, out_size);
+			rv = C_SignFinal(session, out, &out_size);
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err_destr_obj;
+
+			(void)ADBG_EXPECT_BUFFER(c, test->out,
+						 test->out_len,
+						 out, out_size);
+		}
+
+		/* Test 2 step update signature */
+		rv = C_SignInit(session, test->mechanism, key_handle);
+		if (!ADBG_EXPECT_CK_OK(c, rv))
+			goto err_destr_obj;
+
+		if (test->in != NULL) {
+			rv = C_SignUpdate(session,
+					  (void *)test->in, test->in_incr);
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err_destr_obj;
+
+			rv = C_SignUpdate(session,
+					  (void *)(test->in + test->in_incr),
+					  test->in_len - test->in_incr);
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err_destr_obj;
+		}
+
+		out_size = sizeof(out);
+		memset(out, 0, sizeof(out));
+
+		rv = C_SignFinal(session, out, &out_size);
+		if (!ADBG_EXPECT_CK_OK(c, rv))
+			goto err_destr_obj;
+
+		(void)ADBG_EXPECT_BUFFER(c, test->out,
+					 test->out_len, out, out_size);
+
+		/* Test 3 signature in one shot */
+		if (test->in != NULL) {
+			rv = C_SignInit(session, test->mechanism, key_handle);
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err_destr_obj;
+
+			/* Test too short buffer case */
+			out_size = 1;
+			rv = C_Sign(session,(void *)test->in, test->in_len,
+				    out, &out_size);
+			if (!ADBG_EXPECT_CK_RESULT(c, CKR_BUFFER_TOO_SMALL, rv))
+				goto err_destr_obj;
+
+			/*
+			 * Test NULL buffer case with size as 0
+			 * to get the out_size
+			 */
+			out_size = 0;
+			rv = C_Sign(session, (void *)test->in, test->in_len,
+				    NULL, &out_size);
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err_destr_obj;
+
+			/* Get to full output */
+			memset(out, 0, out_size);
+			rv = C_Sign(session,(void *)test->in, test->in_len,
+				    out, &out_size);
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err_destr_obj;
+
+			(void)ADBG_EXPECT_BUFFER(c, test->out,
+						 test->out_len,
+						 out, out_size);
+		}
+
+		rv = C_DestroyObject(session, key_handle);
+		if (!ADBG_EXPECT_CK_OK(c, rv))
+			goto err;
+
+		Do_ADBG_EndSubCase(c, NULL);
+	}
+	goto out;
+
+err_destr_obj:
+	ADBG_EXPECT_CK_OK(c, C_DestroyObject(session, key_handle));
+err:
+	Do_ADBG_EndSubCase(c, NULL);
+out:
+	ADBG_EXPECT_CK_OK(c, C_CloseSession(session));
+err_close_lib:
```

Comment:
nitpicking: these should all be labeled `out_xxx`. It is not an error specific path.

Reply:
`err_destr_obj:` and `err:` are only reached in error path.
`err_close_lib:` will be reached during normal cleanup also, so you have a point, but it's only used explicitly on error.
I'm not sure we have a strict naming convention for this. If there would be a completely separate error path it make sense to use "out" and "err" prefix, but in a case like this where it's mixed we might need to be a bit more flexible.

Reply:
Ok. I find it a (tiny) bit inconsistent to have:
```c
err_foo:
out:
	(...)
err_bar:
	(...)
	return res;
}
```
But fine as is.

Reply:
Feel free to clarify how labels should be named in https://optee.readthedocs.io/en/latest/general/coding_standards.html

Reply:
Kernel coding style docs says "Choose label names which say what the goto does or why the goto exists.". I guess it's enough. I don't want to pollute our coding style with new corner case.
As long as the source implementation in readable, it is fine.

Reply:
Thanks for the clarification. So that means that the labels actually are fully consistent in this case. The key to this is the goto, not the location. Any way, the intent is to make it easier to read the code and I think we've reached that target here.

@ruchi393, sorry for hijacking your PR to discuss coding standards.

Reply:
It's perfectly fine. I was confused myself when choosing the names but what you had suggested seemed readable to me, so I kept them.